File: cpp/Mod_allNodes.cpp
List<Node> helper(Tree t, Node v, List<Node> res) {
    if(isLeaf(t,v)) {
        return res;
    } else {
        for (int i = 0; i < length(children(t,v)); i++) {
            res = append(helper(t, get(children(t,v), i), res), get(children(t,v), i));
        }
        return res;
    }
}

List<Node> allNodes(Tree t) {
    if (isLeaf(t, root(t))) {
        return {root(t)};
    } else {
        return append(helper(t, root(t), {}), root(t));
    }
}

File: hs/Mod_allNodes.hs
module Mod_allNodes ( allNodes ) where
import Prelude hiding (and, or) 
import Debug.Trace
import Splib
import Types
import Mod_subtree

allNodes :: Tree -> [Node]
allNodes tree | size tree == 0 = []
              | isLeaf tree (root tree) = [root tree]
              | otherwise = [root tree] ++ (helper tree [] 0)

helper :: Tree -> [Node] -> Int -> [Node]
helper tree res i | i==length childs = res
                  | otherwise = helper tree (res ++ (allNodes (subtree tree (get childs i)))) (i+1)
                  where
                   childs = children tree (root tree)